### ISSUE TITLE
make autograd.grad at DDP model

### ISSUE THREAD

**Issue by noirmist** [NVIDIA/apex#457]
Can I use torch.autograd.grad between loss and apex DDP model.parameters during training(after wrapping the model with DDP)?

I'm working on implement gradNorm on apex DDP model.
For that, I need to get a gradient between loss and model parameters each iteration.

When I use multi-GPU, my code is stuck at the time when it accesses model parameter.
I'm not sure it is caused by accessing model parameter or autograd.
At using single GPU, everything is working fine except training time. :)

If autograd is not working with apex DDP model (same as torch DDP), 
Or if is there any way to using autograd between loss and model parameter during training time,
Please let me know, Thanks.

**Comment by mcarilli** [comment]
Are you trying to implement a gradient penalty using out of place gradients created via torch.autograd.grad, as shown in [Higher order gradients here](https://github.com/pytorch/pytorch/releases/tag/v0.2.0)?  If so, those out-of-place gradients won't be allreduced, but they don't need to be.  The allreduce will occur when gradients reach the leaves, during your actual `backward()` call.

At this point, `apex.parallel.DistributedDataParallel` is mostly to support internal use cases.  Upstream (`torch.nn.parallel.DistributedDataParallel`) has evolved a lot and its current state is really good.  I recommend trying your code with `torch.nn.parallel.DistributedDataParallel` instead.  The [simple distributed example](https://github.com/NVIDIA/apex/blob/master/examples/simple/distributed/distributed_data_parallel.py#L48-L52) shows how to switch between the two.  Nothing needs to be changed aside from the DDP constructor itself.
